API=PATCH:WorkItem.2 (follow jsonapi spec)

pranavgore09 opened the pull request:
Side effect fixes #466 (How to use - please see [ui#348](https://github.com/almighty/almighty-ui/issues/348) )

WorkItem2 :- Patch the work item. (Allow partial updates & do not change WIT)
Follows `jsonapi spec` to define input payload for PATCH call.

```

{
   "data": {
      "attributes": {
         "system.state": "new",
         "system.title": "Example story",
         "version": "1"
      },
      "id": "42",
      "relationships": {
         "assignee": {
            "data": {
               "id": "6c5610be-30b2-4880-9fec-81e4f8e4fd76",
               "type": "identities"
            }
         },
         "baseType": {
            "data": {
               "id": "system.userstory",
               "type": "workitemtypes"
            }
         }
      },
      "type": "workitems"
   }
}

```
Following attributes are mandatory for every update request with correct value.
1. data.id
2. data.type
3. data.attributes.version

This payload is constructed for updating WI.
Above structure will update work item(if exists) with id 42 and set it's assignee to given UUID if it exists in system.

aslakknutsen commented:
I believe fields == attributes

```json
   "data": {
      "attributes": {
         "system.creator": "user-ref",
         "system.state": "new",
         "system.title": "Example story"
         "version": 5
      },
      "id": "42",
}
```

pranavgore09 commented:
@aslakknutsen , That is possible, but then WIT type and Version will be on the same level of title,description etc, state,  etc. That differs from model definition. I understand jsonapi spec do not say to follow model definition but...
```javascript
// WorkItem represents a work item as it is stored in the database
type WorkItem struct {
	gormsupport.Lifecycle
	ID uint64 `gorm:"primary_key"`
	// Id of the type of this work item
	Type string
	// Version for optimistic concurrency control
	Version int
	// the field values
	Fields Fields `sql:"type:jsonb"`
}

```

aslakknutsen commented:
@pranavgore09 Sure, Model != REST. 

* Version belong more as a ETag on HTTP Header level https://github.com/almighty/almighty-core/issues/155
* Do we even want Version? https://github.com/almighty/almighty-core/issues/420

pranavgore09 commented:
@aslakknutsen @tsmaeder I have addressed above comments except `versioning`. I have not tried `etag` or another similar in this commit. So the new payload looks like -
```jvascript
{
   "data": {
      "attributes": {
         "system.state": "new",
         "system.title": "Example story",
         "type": "system.userstory",
         "version": "1"
      },
      "id": "42",
      "relationships": {
         "assignee": {
            "data": {
               "id": "6c5610be-30b2-4880-9fec-81e4f8e4fd76",
               "type": "identities"
            }
         }
      },
      "type": "workitems"
   }
}
```

As of now. attributes can accept anything. But `version` and `type` are mandatory and validated via our code.

Should we rename `type` to `wiType` in REST payload  ? considering `jsonapi spec` is already having `type` as reserved word kind of.

aslakknutsen commented:
Type should be a relationship.

On Wed, Nov 16, 2016, 08:48 Pranav Gore <notifications@github.com> wrote:

> @aslakknutsen <https://github.com/aslakknutsen> @tsmaeder
> <https://github.com/tsmaeder> I have addressed above comments except
> versioning. I have not tried etag or another similar in this commit. So
> the new payload looks like -
>
> {
>    "data": {
>       "attributes": {
>          "system.state": "new",
>          "system.title": "Example story",
>          "type": "system.userstory",
>          "version": "1"
>       },
>       "id": "42",
>       "relationships": {
>          "assignee": {
>             "data": {
>                "id": "6c5610be-30b2-4880-9fec-81e4f8e4fd76",
>                "type": "identities"
>             }
>          }
>       },
>       "type": "workitems"
>    }
> }
>
> As of now. attributes can accept anything. But version and type are
> mandatory and validated via our code.
>
> Should we rename type to wiType in REST payload ? considering jsonapi spec
> is already having type as reserved word kind of.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/almighty/almighty-core/pull/482#issuecomment-260887949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIEPvwqLVXl7GvLG2VXGiEFBygAKP9jks5q-sNhgaJpZM4Kvta0>
> .
>

pranavgore09 commented:
- `type` is taken out of fields and made it as a `relationship`

```javascript
{
   "data": {
      "attributes": {
         "system.state": "new",
         "system.title": "Example story",
         "version": "1"
      },
      "id": "42",
      "relationships": {
         "assignee": {
            "data": {
               "id": "6c5610be-30b2-4880-9fec-81e4f8e4fd76",
               "type": "identities"
            }
         },
         "baseType": {
            "data": {
               "id": "system.userstory",
               "type": "workitemtypes"
            }
         }
      },
      "type": "workitems"
   }
}

```
ToDo:
- add tests all around patch API

pranavgore09 commented:
@aslakknutsen @tsmaeder : If I go with `etag` for versions, it need to have GET call, so I am keeping `version` as it is for now in `Attributes`, and this PR will do update of WI only. Else it will become big/clumsy. Once this is in master I will start off GET WI using workitem.2 jsonapi

Currently our path URL is - `http://localhost:8080/api/workitems.2/1` in which I am concerned about `id` field in URL.
`jsonapi spec` says that ID should be in `data` along with `type`.
So, we can remove `id` in URL, but then `item exists or not` check will be done in our code and not from `goadesign` ,any suggestions on this ?

sbose78 commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
+	a.Attribute("type", d.String, func() {
+		a.Enum("workitems")
+	})
+	a.Attribute("id", d.String, "ID of the work item which is being updated", func() {
```
Just like in https://github.com/almighty/almighty-core/blob/master/design/user_types.go#L25
we do not include the ID in the payload because we do it allow updating the ID, and that it could be retrieved from the REST URL, I am wondering if we should include it here.

pranavgore09 replied:
yes @sbose78 , you are right, I have mentioned that [here](https://github.com/almighty/almighty-core/pull/482#issuecomment-261141350)
But ID should be in payload as per specifications of jsonapi.org mentioned [here](http://jsonapi.org/format/#document-resource-objects)

sbose78 commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,134 @@
+		uuidStr := assigneeData.ID
+		assigneeUUID, err := uuid.FromString(uuidStr)
+		if err != nil {
+			return nil, NewBadParameterError("data.relationships.assignee.data.id should be UUID", uuidStr)
```
The first argument of the `NewBadParameterError` should be the paramater name. 
The error message would be constructed internally.

pranavgore09 replied:
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/dbc7935d0d707ca9257c79a1e815601e86b511a3

sbose78 commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,134 @@
+		}
+		_, err = identityRepo.Load(ctx, assigneeUUID)
+		if err != nil {
+			return nil, NewBadParameterError("data.relationships.assignee.data.id not found", uuidStr)
```
The first argument of the `NewBadParameterError` should be the paramater name. 
The error message would be constructed internally.

pranavgore09 commented:
@aslakknutsen , should we allow user to update `type` of WI in update API ?
Use case is `change WI from bug to userstory`.
If yes, while updating we need to check for `fields` only, or any other part should get updated ?

In current scenario of PATCH, I have made `type` as a relationship. And it is `required` as of now, which I will be changing.
If provided in payload then we will update type else we will use existing type of the WI and check for fields if they are present or not as described by new type.

pranavgore09 commented:
@aslakknutsen , should we allow user to update `type` of WI in update API ?
Use case is `change WI from bug to userstory`.
If yes, while updating we need to check for `fields` only, or any other part should get updated ?

In current scenario of PATCH, I have made `type` as a relationship. And it is `required` as of now, which I will be changing.
If provided in payload then we will update type else we will use existing type of the WI and check for fields if they are present or not as described by new type.

pranavgore09 commented:
Response for PATCH WI -
```javascript
{
    "data": {
        "attributes": {
            "system.creator": "d8958274-f741-4558-b117-37cb2ed50d09",
            "system.description": null,
            "system.remote_item_id": null,
            "system.state": "open",
            "system.title": "cool title",
            "version": 8
        },
        "id": "1",
        "relationships": {
            "assignee": {
                "data": {
                    "id": "75743e50-641c-4ca5-a10d-9ae824f752f6",
                    "type": "identities"
                }
            },
            "baseType": {
                "data": {
                    "id": "system.userstory",
                    "type": "workitemtypes"
                }
            }
        },
        "type": "workitems"
    },
    "links": {
        "self": "http://localhost:8080/api/workitems.2/1"
    }
}
```

aslakknutsen commented:
Doesn't have to be part of this, but system.creator is a relationship as well.  And technically the relationships should have a link to the target (we don't have a single User service atm, but we do have a Type endpoint)

aslakknutsen commented:
@pranavgore09 Have you looked at this part of the spec? http://jsonapi.org/format/#crud-updating-relationships

pranavgore09 commented:
@aslakknutsen yes and that introduces new endpoint of CRUD API for `assignee relation`.
We decided to use patch for modifying `assignee` and earlier it was part of Attributes then moved to relationship. now this PR makes update api according to jsonapi spec which is not needed to set assignee.
Should I revert all code above ?

pranavgore09 commented:
Just now I have pushed the code for `removing assignee` because it was already done.
But I can revert to include new relationship APIs instead of patch call altogether.

pranavgore09 commented:
@aslakknutsen IMO, as this API is working and integrated [here](https://github.com/sanbornsen/almighty-ui/tree/user_assign). I keep this codebase as it is for updating WI.
And will add relationships UPDATE and DELETE.

baijum commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
 
+// UpdateWorkItemJSONAPIPayload defines top level structure from jsonapi specs
+// visit : http://jsonapi.org/format/#document-top-level
+var UpdateWorkItemJSONAPIPayload = a.Type("UpdateWorkItemJSONAPIPayload", func() {
```
Do we need to export names?

pranavgore09 replied:
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/22a20be904b51c078ea400b03c95b3729230b003

baijum commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
+})
+
+// WorkItemDataForUpdate defines how an update payload will look like
+var WorkItemDataForUpdate = a.Type("WorkItemDataForUpdate", func() {
```
Do we need to export names?

pranavgore09 replied:
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/22a20be904b51c078ea400b03c95b3729230b003

baijum commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
+})
+
+// WorkItemRelationships defines only `assignee` as of now. To be updated
+var WorkItemRelationships = a.Type("WorkItemRelationships", func() {
```
Do we need to export names?

pranavgore09 replied:
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/22a20be904b51c078ea400b03c95b3729230b003

pranavgore09 commented:
@aslakknutsen @baijum 
```javascript
{"data": {"attributes": {"version": "2", "unkown_attribute": "some value"},"id": "2","type": "workitems"}}
```
For input like this, when `unknown` attributes are added in PATCH, that unknown attribute will not be added to DB or anywhere else, but `version` get plus one.

Currently `attributes` is `map[string]interface{}` so no control over it, no code verifies attribute names are valid or not.

Any suggestions ?

tsmaeder commented:
Mmhh...WokItems! Fixed title ;-)

pranavgore09 commented:
Oops, it took me a while to get the difference :P thanks @tsmaeder

kwk commented on design/media_types.go:
```diff
@@ -58,6 +58,23 @@ var workItem = a.MediaType("application/vnd.workitem+json", func() {
+	a.Required("data")
+	a.View("default", func() {
+		a.Attribute("links", func() {
+			a.Attribute("self", d.String)
```
There's no need to repeat the attribute definition here. It's enough to just write `a.Attribute("self")`. The definition will be taken from `WorkItemLinks` just like with the `data` attribute.

pranavgore09 replied:
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/d8c08aea9e5f6cc075bed42bc55fb0972ec34018

kwk commented on design/resources.go:
```diff
@@ -122,6 +122,28 @@ var _ = a.Resource("workitem.2", func() {
+		a.Routing(
+			a.PATCH("/:id"),
+		)
+		a.Description("update the given work item with given id.")
```
The description should be "update the ~~given~~ work item with *the* given id"

pranavgore09 replied:
thanks for correcting, done in https://github.com/almighty/almighty-core/pull/482/commits/d8c08aea9e5f6cc075bed42bc55fb0972ec34018

kwk commented on design/resources.go:
```diff
@@ -122,6 +122,28 @@ var _ = a.Resource("workitem.2", func() {
+			// ToDo update to struct which complies to jsonapi
+			a.Media(workItem2)
+		})
+		a.Response(d.BadRequest, func() {
```
We later need to change the responses like so, once my PR (#421) gets merged:

```go
		a.Response(d.BadRequest, JSONAPIErrors)
		a.Response(d.InternalServerError, JSONAPIErrors)
		a.Response(d.NotFound, JSONAPIErrors)
		a.Response(d.Unauthorized, JSONAPIErrors)
```

This is because all JSONAPI endpoints must/should return this. This is just a reminder.

pranavgore09 replied:
thanks for pointing out, yes will keep in mind. added as a ToDo in https://github.com/almighty/almighty-core/pull/482/commits/d8c08aea9e5f6cc075bed42bc55fb0972ec34018

kwk commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
+	})
+	a.Attribute("relationships", workItemRelationships)
+	// relationships must be required becasue we MUST have workItemType during PATCh
+	a.Required("type", "id", "attributes")
```
I'm not sure that `attributes` has to be a required field. If you just want to update the relationship (e.g. the assignee), then you only need to touch the `relationships` element? What do you think?

pranavgore09 replied:
Thought behind keeping it Required is - "Version" is taken from attributes and that is required field, so while `Update` we must have version.
As mentioned in the [description](https://github.com/almighty/almighty-core/pull/482#issue-188751436) we need "version" to update, and that is in attributes. So that was the idea behind keeping it required.

kwk replied:
Agreed.

kwk commented on design/user_types.go:
```diff
@@ -38,6 +38,71 @@ var UpdateWorkItemPayload = a.Type("UpdateWorkItemPayload", func() {
+	a.Attribute("id", d.String, "UUID of the identity", func() {
+		a.Example("6c5610be-30b2-4880-9fec-81e4f8e4fd76")
+	})
+	a.Required("type")
```
I wonder if `id` isn't a required attribute in this case. What do you do with just the `type` string? See `BaseTypeData` for example, there both attributes are required as I would expect it.

pranavgore09 replied:
As mentioned [here](https://github.com/almighty/almighty-ui/issues/348#issue-191438557), I am using `nil` valud of ID as `remove assignee`. Currently I am not following [this](http://jsonapi.org/format/#crud-updating-relationships)  to update relationships exclusively. These will come later, I have already started work on the same. tracked #513

kwk replied:
ok

kwk commented on design/user_types.go:
```diff
@@ -143,3 +208,10 @@ var identityData = a.Type("IdentityData", func() {
+// WorkItemLinks has `self` as of now according to http://jsonapi.org/format/#fetching-resources
+var WorkItemLinks = a.Type("WorkItemLinks", func() {
+	a.Attribute("self", d.String, func() {
+		a.Example("http://localhost:8080/api/workitems.2/1")
```
Maybe it is better for documentation if the example URL contains `api.almighty.io` rather than `localhost:8080`, don't you think?

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,118 @@
+	log.Printf("updated item to %v\n", newWi)
+	result, err := wiType.ConvertFromModel(newWi)
+	if err != nil {
+		return nil, InternalError{simpleError{err.Error()}}
```
Please use the constructor methods like `NewInternalError` instead of creating error objects directly.

pranavgore09 replied:
thanks, will keep in mind for upcoming work. Fixed in https://github.com/almighty/almighty-core/pull/482/commits/d8c08aea9e5f6cc075bed42bc55fb0972ec34018

kwk commented on workitem_blackbox_test.go:
```diff
@@ -326,6 +329,36 @@ func getWorkItemTestData(t *testing.T) []testSecureAPI {
 		},
+		// adding security tests for workitem.2 endpoint
+		{
+			method:             "PATCH",
```
Please use `http.MethodPatch` instead. the same of the other elements.

pranavgore09 replied:
helpful, replaced all such instances from that file in https://github.com/almighty/almighty-core/pull/482/commits/1c74a7a920d535fcf5e5ef3254b85dbc88b06a7a

kwk commented on workitem_blackbox_test.go:
```diff
@@ -326,6 +329,36 @@ func getWorkItemTestData(t *testing.T) []testSecureAPI {
+		{
+			method:             "PATCH",
+			url:                "/api/workitems.2/12345",
+			expectedStatusCode: 401,
```
Please use `http.StatusUnauthorized` instead of 401.

pranavgore09 replied:
helpful, replaced all such instances from that file in https://github.com/almighty/almighty-core/pull/482/commits/1c74a7a920d535fcf5e5ef3254b85dbc88b06a7a

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+func getMinimumRequiredUpdatePayload(wi *app.WorkItem) *app.UpdateWorkItemJSONAPIPayload {
+	return &app.UpdateWorkItemJSONAPIPayload{
+		Data: &app.WorkItemDataForUpdate{
+			Type: "workitems",
```
I'd suggest to make all `Type`s a string constant.

pranavgore09 replied:
now using constants in https://github.com/almighty/almighty-core/pull/482/commits/97411e3fa46127798b490d59166c4713889a36f0.
thanks.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+
+	// update assignee relationship and verify
+	newUser := createOneRandomUserIdentity(svc.Context, DB)
+	assert.NotNil(t, newUser)
```
You should use `require.NotNil(t, newUser)` because your further testing depends on `newUser` not being `nil`.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+			Type: "identities",
+		},
+	}
+	test.UpdateWorkitem2BadRequest(t, svc.Context, svc, controller2, wi.ID, patchPayload)
```
It would have been nice if each test for a result (e.g. OK or BadRequest) was in its own test function. You can use a test suite to handle common setup and tear down phases for all tests.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+	svc := testsupport.ServiceAsUser("TestUpdateWI2-Service", almtoken.NewManager(pub, priv), account.TestIdentity)
+	assert.NotNil(t, svc)
+	controller := NewWorkitemController(svc, gormapplication.NewGormDB(DB))
+	assert.NotNil(t, controller)
```
Again, please use `require` here instead.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+	minimumPayload.Data.Relationships = &app.WorkItemRelationships{}
+
+	tempUser := createOneRandomUserIdentity(svc.Context, DB)
+	assert.NotNil(t, tempUser)
```
Use`require`.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+	minimumPayload.Data.Relationships.Assignee = assignee
+
+	controller2 := NewWorkitem2Controller(svc, gormapplication.NewGormDB(DB))
+	assert.NotNil(t, controller2)
```
Use `require`.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+	require.NotNil(t, updatedWI)
+	assert.Equal(t, *updatedWI.Data.Relationships.Assignee.Data.ID, tempUserUUID)
+
+	// retrieve work item from Db and verify
```
The update response should be the work item you can use for verification. No need to fetch it again from the DB.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -606,3 +642,433 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+// Removing assignee
+func TestUpdateRemoveAssignee(t *testing.T) {
+	resource.Require(t, resource.Database)
+	pub, _ := almtoken.ParsePublicKey([]byte(almtoken.RSAPublicKey))
```
Please don't repeat so much setup and use a test suite instead. See `workitemtype_blackbox_test.go` for an example.

pranavgore09 commented on design/media_types.go:
```diff
@@ -58,6 +58,23 @@ var workItem = a.MediaType("application/vnd.workitem+json", func() {
+// workItem2 is the media type for work items
+var workItem2 = a.MediaType("application/vnd.workitem2+json", func() {
+	a.TypeName("WorkItem2")
+	a.Description("A work item hold field values according to a given field type in JSONAPI form")
```
thanks, done in https://github.com/almighty/almighty-core/pull/482/commits/d8c08aea9e5f6cc075bed42bc55fb0972ec34018

tsmaeder commented:
This is new API, right? So why have it under /workitem.2 and not /workitem?

pranavgore09 commented:
@tsmaeder yes this is new `Update` action, added to existing `workitem.2` resource.
UI is still using old /workitem update API, so need to keep it as it is until UI migrates to workitem.2 update logic.
As of now, they have started using List action from workitem.2.
Once this PR goes upstream, they will start using workitem.2 for update.

tsmaeder commented:
@pranavgore09 the idea is that the workitem.2 resource will be deprecated once the UI has switched to using the new list API. New API should be added to workitem, now workitem.2. Since we are using a different http method, this is not a breaking change, so no need for using workitem.2.

pranavgore09 commented:
@tsmaeder hmm yes, okay but UI is already using new List method.
If this work is added to `workitem` as a patch action then `workitem` resource will have mixture of non-jsonpai and  jsonapi actions.
Hence IMHO I feel we should keep all jsonapi based apis in `workitem.2` and we will rename it to `workitem` once UI completely shifts to new APIs. I understand UI needs to re-work on URLs but I think it should be easily configurable.  Also putting jsonapi based changes in `workitem.2` makes the resource consistent. Also we will have create/delete actions based on jsonapi spec which will go in `workitem.2`.

kwk commented on design/user_types.go:
```diff
@@ -457,3 +522,10 @@ var RelationWorkItemData = a.Type("RelationWorkItemData", func() {
 })
+
+// WorkItemLinks has `self` as of now according to http://jsonapi.org/format/#fetching-resources
+var WorkItemLinks = a.Type("WorkItemLinks", func() {
```
This name might be misleading because work item links are the once that I implemented in PR #421

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,121 @@
+			return nil, NewBadParameterError("version", version)
+		}
+	} else {
+		return nil, VersionConflictError{simpleError{"version is mandatory"}}
```
Please use `NewVersionConflictError`

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,121 @@
+		return nil, VersionConflictError{simpleError{"version is mandatory"}}
+	}
+	if res.Version != version {
+		return nil, VersionConflictError{simpleError{"version conflict"}}
```
Please use `NewVersionConflictError`

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,121 @@
+
+	rel := wi.Relationships
+	// TODO
+	// if rel.Assignee.Data == nil then remove the relationship for WI
```
So removing an assignee isn't possible at the moment?

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,121 @@
+
+	if err := tx.Save(&newWi).Error; err != nil {
+		log.Print(err.Error())
+		return nil, InternalError{simpleError{err.Error()}}
```
Please use `NewInternalError`

kwk commented on models/workitem2_repository.go:
```diff
@@ -0,0 +1,121 @@
+				return nil, NewBadParameterError("data.relationships.assignee.data.id", uuidStr)
+			}
+			wi.Attributes[SystemAssignee] = *uuidStr
+			//  ToDO : make it a list and append
```
I'd highly suggest to implement it so we can have multiple assingees per WI even though the requirement might not be there yet.

pranavgore09 replied:
Yes I understand that,
But this will extend the scope of task to much extent, because it needs change in `type` of assignee in WI, then change in input, change in assignment logic, and while reading the WorkItem. Plus adding tests for the same. And then review on PR.
So most likely I will end up not doing it complete 100% (merge in master) by tomorrow.
And it is not blocker so....can that be done out of this PR ?
any suggestion @aslakknutsen ?

aslakknutsen replied:
I believe we said under planning that Core should support multi assignee, while UI should only show 1 for now... But Create a new issue to track it. It doesn't have to be part of this.

pranavgore09 replied:
Thanks @aslakknutsen , tracking it in #534 . added that to current sprint. Let me know if that should not be part of sprint.

kwk commented on workitem.2.go:
```diff
@@ -165,3 +168,77 @@ func buildAbsoluteURL(req *goa.RequestData) string {
+
+// ConvertWorkItemToJSONAPI is responsible for converting given WorkItem model object into a
+// response resource object by jsonapi.org specifications
+func (c *Workitem2Controller) ConvertWorkItemToJSONAPI(ctx *app.UpdateWorkitem2Context, wi *app.WorkItem) *app.WorkItem2 {
```
I highly suggest to pass in the second parameter as a value and not as a pointer: `wi app.WorkItem`. The reason is that I ran into a lot of troubles with the pointer that weren't obvious at first. For example when you create the `app.WorkItem2` and assign a pointer value from the `wi`.

kwk commented on workitem.2.go:
```diff
@@ -165,3 +168,77 @@ func buildAbsoluteURL(req *goa.RequestData) string {
+		},
+		Data: &app.WorkItemDataForUpdate{
+			ID:   wi.ID,
+			Type: "workitems",
```
Please use `models.EndpointWorkItems` instead of `"workitems"`.

kwk commented on workitem.2.go:
```diff
@@ -165,3 +168,77 @@ func buildAbsoluteURL(req *goa.RequestData) string {
+				BaseType: &app.RelationshipBaseType{
+					Data: &app.BaseTypeData{
+						ID:   wi.Type,
+						Type: "workitemtypes",
```
Please use `models.EndpointWorkItemLinkTypes` instead of `"workitemtypes"`.

kwk commented on workitem.2.go:
```diff
@@ -165,3 +168,77 @@ func buildAbsoluteURL(req *goa.RequestData) string {
+				op.Data.Relationships.Assignee = &app.RelationAssignee{
+					Data: &app.AssigneeData{
+						ID:   &valStr,
+						Type: "identities",
```
Please create a constant for this:

```go
const(
    EndpointIdentities          = "identities"
)
```

kwk commented on workitem.2.go:
```diff
@@ -165,3 +168,77 @@ func buildAbsoluteURL(req *goa.RequestData) string {
+		}
+		wi, err := appl.WorkItems2().Save(ctx, toSave)
+		if err != nil {
+			switch err := err.(type) {
```
Please rewrite this switch like so to include the correct JSONAPI errors in the response.

```go
switch err := err.(type) {
case models.BadParameterError:
	jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrBadRequest(fmt.Sprintf("Error updating work item: %s", err.Error())))   
	return ctx.BadRequest(jerrors)
case models.NotFoundError:
	jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrNotFound(err.Error()))   
	return ctx.NotFound(jerrors)
case models.VersionConflictError:
	jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrBadRequest(fmt.Sprintf("Error updating work item: %s", err.Error())))   
	return ctx.BadRequest(jerrors)
default:
	log.Printf("Error updating work items: %s", err.Error())
jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInternal(err.Error()))   
	return ctx.InternalServerError(jerrors)
}
```

This will only work if you update the `resources.go` and specify a `JSONAPIErrors` as the second parameter to the error response types:

```go
		a.Response(d.BadRequest, JSONAPIErrors)
		a.Response(d.InternalServerError, JSONAPIErrors)
		a.Response(d.NotFound, JSONAPIErrors)
```

kwk commented on workitem_blackbox_test.go:
```diff
@@ -480,3 +518,288 @@ func TestPagingDefaultAndMaxSize(t *testing.T) {
+	assert.Equal(s.T(), updatedWI.Data.Attributes[models.SystemDescription], modifiedDescription)
+}
+
+func (s *WorkItem2Suite) TestWI2UpdateMultipleScenarios() {
```
Why not put each scenario in a distinct test method?

kwk commented on design/resources.go:
```diff
@@ -110,6 +110,29 @@ var _ = a.Resource("workitem.2", func() {
+			a.Media(workItem2)
+		})
+		//ToDo: Error responses need modifications as per https://github.com/almighty/almighty-core/pull/421
+		a.Response(d.BadRequest, func() {
```
Use `a.Response(d.BadRequest, JSONAPIErrors)` instead.

kwk commented on design/resources.go:
```diff
@@ -110,6 +110,29 @@ var _ = a.Resource("workitem.2", func() {
+		a.Response(d.BadRequest, func() {
+			a.Media(d.ErrorMedia)
+		})
+		a.Response(d.InternalServerError)
```
Use `a.Response(d.InternalServerError, JSONAPIErrors)` instead.

kwk commented on design/resources.go:
```diff
@@ -110,6 +110,29 @@ var _ = a.Resource("workitem.2", func() {
+			a.Media(d.ErrorMedia)
+		})
+		a.Response(d.InternalServerError)
+		a.Response(d.NotFound)
```
Use `a.Response(d.NotFound, JSONAPIErrors)` instead.

kwk commented on design/resources.go:
```diff
@@ -110,6 +110,29 @@ var _ = a.Resource("workitem.2", func() {
+		})
+		a.Response(d.InternalServerError)
+		a.Response(d.NotFound)
+		a.Response(d.Unauthorized)
```
Use `a.Response(d.Unauthorized, JSONAPIErrors)` instead.

kwk commented on design/user_types.go:
```diff
@@ -524,7 +524,7 @@ var RelationWorkItemData = a.Type("RelationWorkItemData", func() {
 })
 
 // WorkItemLinks has `self` as of now according to http://jsonapi.org/format/#fetching-resources
```
The comment doesn't reflect the new name for the type.
